Add proposal for `params Span<T>` (#5382)
dotnet/csharplang: proposals/params-span.md

If we extend `params` types to include the `ref struct` types `Span<T>` and `ReadOnlySpan<T>`, where values of those types cannot escape the call stack, the array at the call site may be created on the stack instead.

And if we're extending `params` to other types, we  could also allow `params IEnumerable<T>` to avoid allocating and copying collections at call sites that have an `IEnumerable<T>` rather than `T[]`.

The benefits of `params ReadOnlySpan<T>` and `params Span<T>` are primarily for new APIs. Existing commonly used APIs such as `Console.WriteLine()` and `StringBuilder.AppendFormat()` already have overloads that avoid array allocations for common cases.
```csharp
public static class Console
{
    public static void WriteLine(string value);
    public static void WriteLine(string format, object arg0);
    public static void WriteLine(string format, object arg0, object arg1);
    public static void WriteLine(string format, object arg0, object arg1, object arg2);
    public static void WriteLine(string format, params object[] arg);
}
```

## Detailed design

### Extending `params`

> `params` parameters will be supported with types `Span<T>`, `ReadOnlySpan<T>`, and `IEnumerable<T>`.

Reviewer: i think we could trivially expand this out to support any interface that Array implements (so IList as well).  This can be nice on the consumption side as it allows params, but also requires anything passed in have a known fixed length (which IEnumerable doesn't provide).

A call in _expanded_ form to a method with a `params T[]` or `params IEnumerable<T>` parameter will result in an array `T[]` allocated on the heap.

> A call in _expanded_ form to a method with a `params ReadOnlySpan<T>` or `params Span<T>` parameter will result in an array `T[]` created on the stack if the `params` array has no more than 8 arguments (an arbitrary limit).

Reviewer: do we want the actual language spec to state the limit?  or do we want to leave it to be a compiler/runtime/user possible heuristic?

Reviewer 2: I am wary of arbitrary boundaries that users don't know when they've crossed. Perhaps this is fine for the "simple expanded call" scenario, where getting the stack allocation is just a bonus, but other users will want to be able to make a call that is "stack allocate this or error", i.e. it won't accidentally cross the threshold as the code evolves. I think this is an important reason to have an "explicit" equivalent to the expanded call.

That said, it occurs to me that this was probably discussed in an LDM that I missed, so this might have all been touched on already..

Author: Created an open issue for heuristics. To explicitly create the span on the stack, we should consider allowing `stackalloc [] { x, y, z }`.

WriteLine(fmt, new[] { x, y, z }); // object[] on the stack for WriteLine(string fmt, ReadOnlySpan<object> args);
```

### Array re-use
The compiler _may_ reuse an implicitly allocated array across multiple uses within a single thread executing a method:
- At the same call-site (within a loop) or
- At distinct call-sites if the lifetime of the spans do not overlap, and the array length is sufficient, and
  - the element types are managed types that are considered identical by the runtime, or
  - the element types are unmanaged types of the same size.

An implicitly allocated array may be reused regardless of whether the array was created on the stack or the heap.

### Lowering implicit allocation
For an array `T[]` of constant length `N`, the compiler will synthesize a `struct` with `N` fields of type `T` where the layout and alignment of the fields matches the alignment of elements in `T[]`.
The array will be represented with a value of the synthesized `struct`, and a `Span<T>` created from a `ref` to the first field.

For example, a call to `Console.WriteLine(fmt, x, y, z);` would be emitted as:
```csharp
[StructLayout(LayoutKind.Sequential)]

> internal struct __ValueArray3<T> { public T Item1, Item2, Item3; };

Reviewer 3: May be worth considering the kind of struct we already generate for [fixed size buffers](https://github.com/dotnet/csharplang/blob/main/spec/unsafe-code.md#fixed-size-buffers). Such struct has a single field with the type of an item, but the layout is stretched out. [sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBsCWyaoA6wwQIB2yANACYgDUAPgAIBMAjALABQDAzAARxkokXjgRwAxjF4BhXgG9OvJbzCYAHhEq9MJKWoDa3ALoBuTgF8gA===)

Reviewer 3: From LDM, turns out the fixed-size buffer approach doesn't work for this scenario (GC tracking wouldn't work).
Fixed-size buffers avoid this today by being limited to simple struct types.
```
public unsafe struct C {
    fixed object x[3]; // error CS1663: Fixed size buffer type must be one of the following: bool, byte, short, int, long, char, sbyte, ushort, uint, ulong, float or double
}
```

Reviewer 4: Is there any need for a struct at all? Why not just use locals directly?

Reviewer 5: Locals are not required to be stored in a sequentual order - usually they are not even stored in memory but in CPU register.
I think we should not force a lowering strategy (e.g. stackalloc, custom type with seq. fields or fixed size buffer).
We should just mention those as possible implementations.

var values = new __ValueArray3<object>() { Item1 = x, Item2 = y, Item3 = z };
var span = MemoryMarshal.CreateSpan(ref values.Item1, 3);
Console.WriteLine(fmt, (ReadOnlySpan<object>)span); // WriteLine(string format, params ReadOnlySpan<object?> arg)
```

The synthesized `struct` type will be `internal`, with an unspeakable name.

Since `T` is a valid type argument (and used in `Span<T>`), the synthesized type will be a constructed generic type where the underlying generic type is shared across other call sites in the compilation with the same number of arguments.

> 

Reviewer: it's possible today to have a `params int*[]`.  Whatever we do, we have to make sure that that still works. i bring this up as pointers cannot be type parameters.  so while we may not allow a params span of pointers, we still need to make sure anything we do doesn't break the standard array case.

Author: `params` parameters for types that are not valid type arguments (such as `int*`) will need to continue using `params T[]`.

## Open issues
### Is `params Span<T>` necessary?

> Is there a reason to support `params` parameters of type `Span<T>` in addition to `ReadOnlySpan<T>`? Is allowing mutation within the `params` method useful?

Reviewer: yes.  i think we would want this.  we have params arrays today, and those are mutable.  it would be unfortuante for those apis that depend on this to not be able to use params span and get the benefits.

Author: When might it be useful to modify the `params` array or `Span<T>` within the `params` method? If the parameter is marked `params`, presumably many callers will use _expanded_ form (with an implicit array or span), so the modifications to the array will be ignored.

Reviewer 6: Maybe:

```cs
for (var i = 0; i < someParams.Length; i++)
    someParams[i] ??= string.Empty; // Or something
```

### Is `params IEnumerable<T>` necessary?
If the compiler allows `params ReadOnlySpan<T>`, then new APIs that require `params` could use `params ReadOnlySpan<T>` instead of `params T[]` because `T[]` is implicitly convertible to `ReadOnlySpan<T>`. And even legacy APIs could add a `params ReadOnlySpan<T>` overload where the existing `params T[]` simply delegates to the new overload.

There is no conversion from `IEnumerable<T>` to `ReadOnlySpan<T>` however, so allowing `params IEnumerable<T>` is essentially asking APIs to provide two overloads for `params` methods: `params ReadOnlySpan<T>` and `params IEnumerable<T>`. 

> Are scenarios for `params IEnumerable<T>` sufficiently compelling to justify that?

Reviewer: yes, i believe so.  This feature is not jsut about perf, but about ergonomics.  The requirement to take in an array has been very limiting to many apis, requiring an array overload (for the ease of passing values), and an IEnumerable/IList/etc. overload (to allow other collections to be passed in).   It's been a very longstanding request from teh community that this sort of common pattern not require multiple overloads when one would be clear and sufficient.

Reviewer 4: Could the compiler generate a static instance of an ImmutableArray<T> and share it across all calls to `params IEnumerable<T>`?

Reviewer 7: No because in the case of anything but `Span<T>` type we must assume the implementation smuggles away the argument. Hence re-use is not safe.

Reviewer 4: But if the backing type is an ImmutableArray that shouldn't matter? I'm taking about a case where the params are constants.

Reviewer 7: In that specific case yes we could put it in an `ImmutableArray<T>`. I'm not sure it's worth it. 

- Forces the compiler to depend on a new type and assign specific semantics to it. Generally we only do that for special types. 
- It's not a total win, there is still allocations at the call site (boxing the `ImmutableArray<T>`) and at the callee (allocation of `IEnumerator` during enumeration). 

Don't think we'll save enough for it to be worth the complexity.

Reviewer 4: You called store the ImmutableArray already boxed as an IEnunerable to save on the first allocation. But agreed, less than perfect.

### Explicit `stackalloc`
Should we allow explicit stack allocation of arrays of managed types with `stackalloc` as well?
```csharp
public static ImmutableArray<TResult> Select<TSource, TResult>(this ImmutableArray<TSource> source, Func<TSource, TResult> map)
{
    int n = source.Length;
    Span<TResult> result = n <= 16 ? stackalloc TResult[n] : new TResult[n];
    for (int i = 0; i < n; i++)
        result[i] = map(source[i]);

>     return ImmutableArray.Create(result); // requires ImmutableArray.Create<T>(ReadOnlySpan<T>)

Reviewer 2: I think this requires the 'DoesNotEscape' annotation on ImmutableArray.Create.

Reviewer 5: stackalloc is quite dangerous because: 

> stackalloc expression allocates memory from the current stack frame, and that memory may not be released until the current method call returns. If stackalloc is used in a loop, it can lead to stack overflows due to exhausting the stack memory.

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2014

> *  Otherwise if one member is a non-lifted operator and  the other is a lifted operator, the non-lifted one is better.
> *  Otherwise, neither function member is better.

### Array creation expressions
Array creation expressions that are target-typed to `ReadOnlySpan<T>` or `Span<T>` will be created on the stack if the length of the array is a constant value no more than 8 (an arbitrary limit).
Otherwise the array will be allocated on the heap.

```csharp
Span<int> s = new[] { i, j, k };   // int[] on the stack
WriteLine(fmt, new[] { x, y, z }); // object[] on the stack for WriteLine(string fmt, ReadOnlySpan<object> args);
```

### Array re-use
The compiler _may_ reuse an implicitly allocated array across multiple uses within a single thread executing a method:
- At the same call-site (within a loop) or
- At distinct call-sites if the lifetime of the spans do not overlap, and the array length is sufficient, and
  - the element types are managed types that are considered identical by the runtime, or
  - the element types are unmanaged types of the same size.

An implicitly allocated array may be reused regardless of whether the array was created on the stack or the heap.

> ### Lowering implicit allocation

Reviewer 7: I would consider moving this to an open issues section. Essentially here is one potential way we can lower. 

For the current design I would say that the compiler is going to lower the calls to the most efficient way possible. Specifically avoiding heap allocations when possible. The exact details are still to be determined and may differ based on the target framework and hence the tools available.

Console.WriteLine(fmt, (ReadOnlySpan<object>)span); // WriteLine(string format, params ReadOnlySpan<object?> arg)
```

## Open issues
### Is `params Span<T>` necessary?
Is there a reason to support `params` parameters of type `Span<T>` in addition to `ReadOnlySpan<T>`? Is allowing mutation within the `params` method useful?

### Is `params IEnumerable<T>` necessary?
If the compiler allows `params ReadOnlySpan<T>`, then new APIs that require `params` could use `params ReadOnlySpan<T>` instead of `params T[]` because `T[]` is implicitly convertible to `ReadOnlySpan<T>`. And even legacy APIs could add a `params ReadOnlySpan<T>` overload where the existing `params T[]` simply delegates to the new overload.

There is no conversion from `IEnumerable<T>` to `ReadOnlySpan<T>` however, so allowing `params IEnumerable<T>` is essentially asking APIs to provide two overloads for `params` methods: `params ReadOnlySpan<T>` and `params IEnumerable<T>`. 

Are scenarios for `params IEnumerable<T>` sufficiently compelling to justify that?

### Heuristics
The proposed approach falls back to heap allocation if the array length exceeds an arbitrary limit.

> Do we need a limit, and if so, what are the heuristics?

Reviewer 7: I would generalize this a bit. Rather than recommending specific heuristics I would say that the language, using only the tools we have today, is likely to have to resort to heuristics. Experimentation should establish the limits we agree on.

### Explicit `stackalloc`
Should we allow explicit stack allocation of arrays of managed types with `stackalloc` as well?
```csharp
public static ImmutableArray<TResult> Select<TSource, TResult>(this ImmutableArray<TSource> source, Func<TSource, TResult> map)
{
    int n = source.Length;
    Span<TResult> result = n <= 16 ? stackalloc TResult[n] : new TResult[n];
    for (int i = 0; i < n; i++)
        result[i] = map(source[i]);
    return ImmutableArray.Create(result); // requires ImmutableArray.Create<T>([DoesNotEscape] ReadOnlySpan<T> items)
}
```

This would require runtime support for stack allocation of arrays of non-constant length and any type, and GC tracking of the elements.

> Direct runtime support for stack allocation of arrays of managed types might be useful for lowering implicit allocation as well.

Reviewer 7: The other downside we should discuss here is that the JIT not track the lifetime of  `stackalloc` locals the same way it tracks normal locals. That is because we will immediately grab a `ref` to their contents and that breaks JIT ability to track them. It essentially gives up and considers them always alive. 

That means in order to not introduce inadvertent memory leaks the compiler has to consider manual management strategies. For example when we exit a scope where we did a `stackalloc` we need to consider calling `.Clear()` on the `Span<T>`. 

These problems, and the likely solutions, mirror the problems we have with lifted locals in state machines. Unless we take compensating actions the lifetime of locals won't match the intuition of developers.

# `params Span<T>` and implicit allocation of arrays

## Summary

> Avoid heap allocation for implicit allocation of arrays in specific scenarios such as `params` arguments.

Reviewer 3: ```suggestion
Avoid heap allocation for implicit allocation of arrays in specific scenarios with `params` arguments.
```

Reviewer 3: The motivation for suggestion is that some `params Span<T>` scenarios are likely to still allocate.

## Motivation
`params` array parameters provide a convenient way to call a method that takes an arbitrary length list of arguments.
However, using an array type for the parameter means the compiler must implicitly allocate an array on the heap at each call site.

If we extend `params` types to include the `ref struct` types `Span<T>` and `ReadOnlySpan<T>`, where values of those types cannot escape the call stack, the array at the call site may be created on the stack instead.

And if we're extending `params` to other types, we  could also allow `params IEnumerable<T>` to avoid allocating and copying collections at call sites that have an `IEnumerable<T>` rather than `T[]`.

> The benefits of `params ReadOnlySpan<T>` and `params Span<T>` are primarily for new APIs. Existing commonly used APIs such as `Console.WriteLine()` and `StringBuilder.AppendFormat()` already have overloads that avoid array allocations for common cases.

Reviewer 3: >The benefits [](http://example.com/codeflow?start=0&length=12)

There's also a benefit for existing APIs: they can be re-written to call a single overload.

Author: Existing overloads would need to be retained for backward compatibility however, so a `params ReadOnlySpan<T>` overload could be added, but would not replace existing overloads.

```csharp
public static class Console
{
    public static void WriteLine(string value);
    public static void WriteLine(string format, object arg0);
    public static void WriteLine(string format, object arg0, object arg1);
    public static void WriteLine(string format, object arg0, object arg1, object arg2);
    public static void WriteLine(string format, params object[] arg);
}
```

## Detailed design

### Extending `params`
`params` parameters will be supported with types `Span<T>`, `ReadOnlySpan<T>`, and `IEnumerable<T>`.

> A call in _expanded_ form to a method with a `params T[]` or `params IEnumerable<T>` parameter will result in an array `T[]` allocated on the heap.

Reviewer 3: >_expanded_ form [](http://example.com/codeflow?start=10&length=15)

1. Consider including a link to spec section that defines "expanded form" [https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#applicable-function-member](https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#applicable-function-member)
2. Is allocating a `T[]` array on the heap the only option for `params IEnumerable<T>` in expanded form?

Author: I've added a link for _expanded form_, thanks.

For `params IEnumerable<T>`, the collection will need to be allocated on the heap since the method being called may hold onto a reference to the collection. Presumably it could be an `IEnumerable<T>` other than `T[]` though.

The guarantee the compiler gives is the span will be the expected size and will contain the expected items at any point in user code.

## Open issues
### Is `params Span<T>` necessary?
Is there a reason to support `params` parameters of type `Span<T>` in addition to `ReadOnlySpan<T>`? Is allowing mutation within the `params` method useful?

### Is `params IEnumerable<T>` necessary?
If the compiler allows `params ReadOnlySpan<T>`, then new APIs that require `params` could use `params ReadOnlySpan<T>` instead of `params T[]` because `T[]` is implicitly convertible to `ReadOnlySpan<T>`. And even legacy APIs could add a `params ReadOnlySpan<T>` overload where the existing `params T[]` simply delegates to the new overload.

There is no conversion from `IEnumerable<T>` to `ReadOnlySpan<T>` however, so allowing `params IEnumerable<T>` is essentially asking APIs to provide two overloads for `params` methods: `params ReadOnlySpan<T>` and `params IEnumerable<T>`. 

Are scenarios for `params IEnumerable<T>` sufficiently compelling to justify that?

### Array limits
The compiler may use heuristics to determine when to fallback to heap allocation for the underlying data for spans.

> Experimentation should establish the limits we agree on.

Reviewer 3: As I mentioned in LDM, I'm not convinced that the compiler can determine what is right/preferrable. Maybe we should pick a default (for example some stack allocation) and have an escape for users who prefer the other mode (use explicit array allocation).

Author: Changed the comment to "If heuristics are necessary, ..."
